### DISCUSSION TITLE
Feature Request: Persistent Notification w/ Web Push API

### DISCUSSION THREAD

**Discussion by Warr1024** [jcorporation/myMPD#590] [Ideas]
I would really love to have a persistent notification (especially on Mobile) on each device which:
- Is always present when mpd is playing.
- Displays the current song information.
- Allows me to quickly open myMPD.
- Possibly has other quick actions, like play/pause/stop/next/prev.

It would also be nice to be able to have the option to receive notifications via Web Notifications even when the app is not actually open and running.  This capability may be necessary for the above persistent notification feature, too, since on mobile, PWAs may be killed by the system when in background and need to be externally revived.

Using the Web Push API, the myMPD server can push events to clients and their ServiceWorkers will be spun up to respond to them, which can include displaying/updating Web Notifications.  Clients would need to register themselves with the server, which would then need to persistently store a set of client registrations, and then the server would need to watch for events, and for each one, check the list of registered clients and their filters and send that event to each one.

Currently, the ability to keep me notified in the background is the only reason I use a native mobile app for mpd control instead of myMPD (and the native app is not very reliable either).  Expanding notification use would allow me to use myMPD as the only front-end for mpd across all my devices/platforms.

**Comment by jcorporation** [comment]
Quick actions are implemented in the v8.2.0 branch: #490 

I must look at the API specification. How widely is the support on mobile plattforms? iOS, Android?

**Comment by jcorporation** [comment]
The web push api is based on http/2 server push: 
- https://serviceworke.rs/push-get-payload_server_doc.html
- https://datatracker.ietf.org/doc/html/draft-ietf-webpush-protocol

The integrated mongoose webserver supports only http/1.1. The autor of mongoose don‘t plan to support http/2: https://github.com/cesanta/mongoose/issues/1186

Finaly myMPD could implement this feature request only by changing the embedded webserver. I do not even know if there is a c based webserver library around that supports http/2 and the webpush protocol.

**Reply by Warr1024** [reply]
That's odd, I never noticed any mention of http2 when I tried it out myself, I just built my proof of concept using nodejs express, which I'm pretty sure only does http1.1 as well.

In any event, if h2 is required, it can be provided by a reverse proxy, which is how my own home setup works.

**Comment by jcorporation** [comment]
Here’s an article how that works: https://medium.com/zettle-engineering/beginners-guide-to-web-push-notifications-using-service-workers-cb3474a17679

How I understood:
- The web push service must not be implemented by the application. 
- Each browser connects to the push service of his vendor
- Browser registers this endpoint in the backend app
- Backend app sends notifications to the browser vendors push service
- Browser gets the notifications from this push service 

This works only with working internet connection and not standalone.

I don’t know if this could be implemented without the usage of the browser vendors push service.

**Reply by Warr1024** [reply]
Yes, that's a limitation.  It's controlled by the browser, which chooses its own provider for push delivery.  That provider will be able to see who is sending messages to whom, at what time, and how large the payloads are.  For example, Google will have access to this information for Chrome users.

Users will need to provide informed consent to enable this, so some kind of warning may be appropriate.  Presumably we Android users are already used to Google handling all of our push notifications across most apps, so one more makes little difference; the fact that the payloads are encrypted does help, at least. Other users may want to opt out, but you would need to support users who are unable to use this API anyway.

**Reply by jcorporation** [reply]
In the moment the notifications are handled by a direct websocket connection. 

This proposal could be only a second optional way for notifications. It could be implemented in the future but not in the next planed releases. There are waiting so many other enhancements for implementation.
